src/ui/public/filter_bar/lib/multi_filter.js

```diff
@@ -0,0 +1,93 @@
+import _ from 'lodash';
+
+/**
+ * Create single filter per data field. Fields with multiple filters are merged into a bool query.
+ * Filters without meta.key can not be merged because field is unknown.
+ *
+ * @param {array} filters to merge.
+ * @returns {array} merged filters
+ */
+export function mergeFilters(filters) {
+  // Group filters by key
+  const filterGroups = new Map();
+  const filtersWithNoKey = [];
+  filters.forEach((filter) => {
+    const key = _.get(filter, 'meta.key');
+    if (!key) {
+      filtersWithNoKey.push(filter);
+      return;
+    }
+
+    if (!filterGroups.has(key)) {
+      filterGroups.set(key, [filter]);
+    } else {
+      filterGroups.set(key, filterGroups.get(key).concat(filter));
+    }
+  });
+
+  const filtersWithKey = [];
+  filterGroups.forEach((filterGroup, key) => {
+    if (filterGroup.length === 1) {
+      // Key with only one filter
+      filtersWithKey.push(filterGroup[0]);
+    } else {
+      // Key with multiple filters - combine into OR query
+      const values = [];
+      const shouldFilters = [];
+      const mustNotFilters = [];
+      filterGroup.forEach((filter) => {
+        const value = _.get(filter, 'meta.value', '');
+        if (_.get(filter, 'meta.negate', false)) {
+          values.push('!' + value);
+          mustNotFilters.push(cleanFilter(filter));
+        } else {
+          values.push(value);
+          shouldFilters.push(cleanFilter(filter));
+        }
+      });
+      filtersWithKey.push({
+        bool: {
+          should: shouldFilters,
+          must_not: mustNotFilters
+        },
+        meta: {
+          alias: `${key}: ${values.join()}`
+        }
+      });
```

Comment:
I think we need a better way to do ORs than just cramming them all into a custom filter. When a user goes to edit this filter they'll just get the JSON editor. Ideally they could use the filter editor UI to edit each one of the sub-filters individually. Otherwise ORs are a second class citizen compared to ANDs.

Initially I thought you just needed to use the [phrases filter constructor](https://github.com/elastic/kibana/blob/013b7943dc8358f29c6e52dbefad02b9cefe8fc6/src/ui/public/filter_manager/lib/phrases.js#L3) but what you're doing here is more generic. We probably need to implement ORs as a more fundamental part of the filter bar. This work could really grow legs because we're talking about adding an entirely new concept (grouped, nested queries with boolean logic) to the filter bar.

Alternatively, we could try implementing this in the new kuery language first where we already have first class support for OR. This is one of the main limitations of the filter bar that we were trying to address with the new query language.

Reply:
> When a user goes to edit this filter they'll just get the JSON editor.

@Bargs - I think this is not the case any more now that the new Edit Filter dialog @lukasolson worked on is submitted?  Now it could show something like: 
![screen shot 2017-09-26 at 9 22 30 am](https://user-images.githubusercontent.com/16563603/30862556-736d057e-a29c-11e7-92a7-1c6379865f8d.png)

For the Kuery version, I think we would need a visual indication of unapplied vs applied queries, which would probably be something nice anyway.  Otherwise if you type in a query, walk away from the keyboard without hitting enter, you come back later and don't know if what you are seeing in the visualizations matches what you are seeing in the query bar.

Reply:
@Stacey-Gammon the `is one of` option your screenshot shows creates a [phrases filter](https://github.com/elastic/kibana/blob/013b7943dc8358f29c6e52dbefad02b9cefe8fc6/src/ui/public/filter_manager/lib/phrases.js#L3) which produces a bool query containing only `match_phrase` queries in a single `should` clause. The code in this PR groups together all filters targeting a given field which could include much more than just `match_phrase` queries. It also includes negative filters. So this code is attempting something more complex than `is one of`.